### PR TITLE
Adding Pytest Tests

### PR DESCRIPTION
**Pull request by callum-mcdata**
### Description
This PR adds `pytest` tests. It does NOT implement them as part of CI. That will take place in a following PR so that people don't have to review one massive PR and its instead 1 massive PR and one tiny PR. Wait....

#### Test Types Added:
- **base_metric_types:** ensures that all the base metric types are working
- **compilation**: ensures that metrics show up in manifest. More of a canary for something going wrong in dbt-core
- **expression_metric_types:** ensures that expression metrics are working. We test for a simple calculation AND a ratio calculation to ensure div/0 not reintroduced
- **invalid configs**: ensures that known issues or anti-patterns aren't breaking things
- **metric_options**:
  - **date_grains**: checks that the default date grains of day, week, month all work correctly
  - **dimensions**: ensures that single and multi dimensions provided work correctly
  - **end_date**: ensures that the end date functionality works
  - **multiple_metrics:** ensures that multiple metrics work
  - **start_date:** ensures that the start date functionality works
  - **secondary_calculations:**
    - **period_over_period**: ensures that both period over period options work
    - **period_to_date**: ensures that all allowed options work for all metric types
    - **rolling**: ensures that all allowed options work for all metric types

### CODE REVIEW

File: dev-requirements.txt [hunk 1]
```diff
@@ -1,4 +1,17 @@
...
+# git+https://github.com/dbt-labs/dbt-redshift.git
+# git+https://github.com/dbt-labs/dbt-snowflake.git
+# git+https://github.com/dbt-labs/dbt-bigquery.git
+
+# Most recent release candidates
+dbt-tests-adapter==1.2.0rc1
+dbt-core==1.2.0rc1
+dbt-redshift==1.2.0rc1
+dbt-snowflake==1.2.0rc1
+dbt-bigquery==1.2.0rc1
```
Comment (joellabes) [inline dev-requirements.txt:17]: How will these stay up to date over time?

File: integration_tests/models/materialized_models/fact_orders.sql [hunk 2]
```diff
@@ -1,4 +1,4 @@
 select 
     *
-    ,round(order_total - (order_total/2)) as discount_total
+    ,1 as discount_total
```
Comment (joellabes) [inline integration_tests/models/materialized_models/fact_orders.sql:3]: ❓ this is weird, shouldn't it be a percentage or something? I just spent a bunch of time digging through the files trying to work out why the average discount was always 1 in your expectation seed. 

The thing that worries me here is that if your source data isn't that varied then you don't actually get the benefit of testing the aggregations - what if something has gone horribly wrong and you're actually calculating max/min or something?

File: integration_tests/seeds/expected/base_average_metric__expected.csv [hunk 3]
```diff
@@ -1,303 +1,5 @@
...
-2022-05-31,TRUE,31.000000
-2022-03-27,FALSE,0.000000
-2022-04-11,TRUE,29.000000
-2022-05-20,FALSE,19.600000
\ No newline at end of file
+﻿date_month,had_discount,base_average_metric
+2022-01-01,TRUE,1.00000000000000000000
+2022-01-01,FALSE,1.00000000000000000000
+2022-02-01,FALSE,1.00000000000000000000
+2022-02-01,TRUE,1.00000000000000000000
```
Comment (joellabes) [inline integration_tests/seeds/expected/base_average_metric__expected.csv:5]: This isn't testing much of anything (as mentioned above)

File: macros/secondary_calculations/secondary_calculation_period_over_period.sql [hunk 4]
```diff
@@ -28,7 +28,7 @@
 {% endmacro %}
 
 {% macro default__metric_comparison_strategy_ratio(metric_name, calc_sql) %}
-    cast(coalesce({{ metric_name }}, 0) / nullif({{ calc_sql }}, 0) as {{ dbt_utils.type_float() }})
+    cast(coalesce({{ metric_name }}, 0) as {{ dbt_utils.type_float() }}) / nullif({{ calc_sql }}, 0)
```
Comment (joellabes) [inline macros/secondary_calculations/secondary_calculation_period_over_period.sql:31]: 👍

File: tests/functional/base_metric_types/test_base_average_metric.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      - had_discount
+      - order_country
+"""
+
+# seeds/base_average_metric__expected.csv
+base_average_metric__expected_csv = """
+date_month,base_average_metric
+2022-01-01,1.00000000000000000000
+2022-02-01,1.3333333333333333
+""".lstrip()
```
Comment (joellabes) [inline tests/functional/base_metric_types/test_base_average_metric.py:49]: Okkkkkk so am I understanding this correctly: 

These lines of hardcoded logic are the same as the ones that are up above as proper .sql/.yml/.csv files? So we're duplicating everything? How do they stay in sync? And why do we need the former if it's all just turning into string constants down here? 

idk if this is pythonic or not, but I would strongly prefer something like 

```python
#https://stackoverflow.com/a/49564464/14007029
from pathlib import Path
base_average_metric_sql = Path('../models/base_average_metric.sql').read_text();
base_average_metric_yml = Path('../models/base_average_metric.yml').read_text();
```

Such that we don't have to redefine everything from scratch in two places 🤢

File: tests/functional/base_metric_types/test_base_average_metric.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        results = run_dbt(["seed"])
+        assert len(results) == 2
+
+        # initial run
+        results = run_dbt(["run"])
+        assert len(results) == 3
+
+        # test tests
+        results = run_dbt(["test"]) # expect passing test
+        assert len(results) == 1
```
Comment (joellabes) [inline tests/functional/base_metric_types/test_base_average_metric.py:103]: What is the `results` object here? Is it an array of passing tests? I would sorta prefer that it returned all results, and a pass/fail, and then we could do whatever the python version of `results.all(result => result.statuscode == 'PASS')` is

Comment (joellabes) [reply tests/functional/base_metric_types/test_base_average_metric.py:103]: Because otherwise anytime we add new tests, we have to go through and update the expected number of passes. (or, worse, we add a test which fails, we forget to update the expected count, and it sits there failing forever).

File: tests/functional/base_metric_types/test_base_average_metric.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+        # # # validate that the results include pass
+        result_statuses = sorted(r.status for r in results)
+        assert result_statuses == ["pass"]
```
Comment (joellabes) [inline tests/functional/base_metric_types/test_base_average_metric.py:107]: OK maybe that's what this line is? but then I don't know what the above stuff is, and I still feel very uncomfortable about hardcoded numbers of things in test files

**Review by joellabes — CHANGES_REQUESTED** [review]
I have reviewed five files! I think that is a representative set of the whole, but if there are things you think I should have looked at that got radio silence then let me know. 

It's very possible that some of my complaints here come from being unfamiliar with pytest, but the things that are in here _feel_ very fragile to me.

File: dev-requirements.txt (hunk 1 above)
Comment (callum-mcdata) [reply dev-requirements.txt:17]: We'll manually update these based on the version that we want to test. Right now its the RC candidate but we'll update it to v1.2 once it is no longer RC.

File: tests/functional/base_metric_types/test_base_average_metric.py (hunk 5 above)
Comment (callum-mcdata) [reply tests/functional/base_metric_types/test_base_average_metric.py:49]: Not quite. There are static datasets/definitions contained in `fixtures.py` and those are then imported. The reason that I keep some components outside of it (like the definition of the metric) is so we can understand what metric is being tested while just looking at the single file. 

If we're trying to be more pythonic then I should move that definition into the fixtures file.

File: tests/functional/base_metric_types/test_base_average_metric.py (hunk 6 above)
Comment (callum-mcdata) [reply tests/functional/base_metric_types/test_base_average_metric.py:103]: @dbeatty10 would have more context around this area. I suspect it returns the results of each command and then we confirm that it matches what we expect. So if we added a new test we'd want to assert that only two tests were running

File: integration_tests/models/materialized_models/fact_orders.sql (hunk 2 above)
Comment (callum-mcdata) [reply integration_tests/models/materialized_models/fact_orders.sql:3]: Very good point - I was trying to keep it as simple as possible but if it doesn't represent the actual calculation then thats not helpful

File: tests/functional/base_metric_types/test_base_average_metric.py (hunk 7 above)
Comment (dbeatty10) [reply tests/functional/base_metric_types/test_base_average_metric.py:107]: The example containing hard-coded numbers comes from [here](https://docs.getdbt.com/docs/contributing/testing-a-new-adapter#example-a-simple-test-case).

`dbt_metrics` doesn't need to do it that way if it doesn't want to. The important part is to verify that it works as expected for all the possible edge cases. And that it fails in the ways expected with bad input.

File: tests/functional/base_metric_types/test_base_average_metric.py (hunk 6 above)
Comment (dbeatty10) [reply tests/functional/base_metric_types/test_base_average_metric.py:103]: The example containing hard-coded numbers comes from [here](https://docs.getdbt.com/docs/contributing/testing-a-new-adapter#example-a-simple-test-case).

`dbt_metrics` doesn't need to do it that way if it doesn't want to. The important part is to verify that it works as expected for all the possible edge cases. And that it fails in the ways expected with bad input.

File: integration_tests/models/materialized_models/fact_orders.sql (hunk 2 above)
Comment (callum-mcdata) [reply integration_tests/models/materialized_models/fact_orders.sql:3]: Oh actually this is still a holdover from integration testing! These will mainly just be used for local development as we move the CI testing to pytest to test functionality. So the old version with round is retained in the pytest models 👯

File: integration_tests/seeds/expected/base_average_metric__expected.csv (hunk 3 above)
Comment (callum-mcdata) [reply integration_tests/seeds/expected/base_average_metric__expected.csv:5]: This is fixed in the pytest framework - everything within the integration tests folder is a holdover from when we were doing it that way and I'm hesitant to delete anything until we've got it all migrated over.

File: tests/functional/base_metric_types/test_base_average_metric.py (hunk 5 above)
Comment (joellabes) [reply tests/functional/base_metric_types/test_base_average_metric.py]: OK! I think I was caught up on thinking that the stuff in the integration tests project was this stuff's twin. Carry on!
